### PR TITLE
GLTFLoader: Implement KHR_materials_clearcoat.

### LINKED ISSUE #18677: MeshPhysicalMaterial: Support clearcoatMap, clearcoatRoughnessMap
MeshPhysicalMaterial currently offers:

- clearcoat
- clearcoatRoughness
- clearcoatNormalMap

The [glTF extension `KHR_materials_clearcoat`](https://github.com/KhronosGroup/glTF/pull/1756) will also require texture versions of the first two, which I'd propose adding to (only) MeshPhysicalMaterial:

- clearcoatMap
- clearcoatRoughnessMap

### PR DESCRIPTION
**Pull request by donmccurdy**
Clearcoat: without (left) and with (right).

Screen Shot 2020-02-19 at 8 20 03 PM

To do:
- [x] ~~Extension may need to be read from the material def, not the pbrSpecularGlossiness sub-object.~~
- [x] ~~Test against more cases, e.g. https://github.com/KhronosGroup/glTF-Sample-Models/pull/251.~~

### CODE REVIEW

File: src/renderers/shaders/ShaderChunk/lights_physical_fragment.glsl.js [hunk 1]
```diff
@@ -21,12 +21,28 @@ material.specularRoughness = min( material.specularRoughness, 1.0 );
...
+
+  #ifdef USE_CLEARCOATMAP
+
+    material.clearcoat *= texture2D( clearcoatMap, vUv ).x;
+
+  #endif
+
+  #ifdef USE_CLEARCOAT_ROUGHNESSMAP
+
+    material.clearcoatRoughness *= texture2D( clearcoatRoughnessMap, vUv ).x;
```
Comment (takahirox) [inline src/renderers/shaders/ShaderChunk/lights_physical_fragment.glsl.js]: In the [current gltf extension proposal](https://github.com/KhronosGroup/glTF/pull/1756), the values of clearcoat roughness texture seems sampled from G (Y) channel, not R(X).

Comment (donmccurdy) [reply src/renderers/shaders/ShaderChunk/lights_physical_fragment.glsl.js]: Ahh good catch, thanks. Let's discuss there, I'm not sure. https://github.com/KhronosGroup/glTF/pull/1756/files#r381750517

Comment (donmccurdy) [reply src/renderers/shaders/ShaderChunk/lights_physical_fragment.glsl.js]: Updated. ✅

**Comment by donmccurdy** [comment]
Successfully rendering @emackey's excellent test asset now:

Screen Shot 2020-02-20 at 11 50 16 PM

File: src/renderers/shaders/ShaderLib.js [hunk 2]
```diff
@@ -299,12 +299,15 @@ ShaderLib.physical = {
...
 		{
-			transparency: { value: 0 },
 			clearcoat: { value: 0 },
+			clearcoatMap: { value: null },
 			clearcoatRoughness: { value: 0 },
-			sheen: { value: new Color( 0x000000 ) },
+			clearcoatRoughnessMap: { value: null },
+			clearcoatNormalMap: { value: null },
 			clearcoatNormalScale: { value: new Vector2( 1, 1 ) },
 			clearcoatNormalMap: { value: null },
```
Comment (moraxy) [inline src/renderers/shaders/ShaderLib.js]: `google-closure-compiler` complains about `clearcoatNormalMap` being used twice, failing the build
```
build/three.js:15412: ERROR - [JSC_DUPLICATE_OBJECT_KEY] Object literal contains illegal duplicate key "clearcoatNormalMap", disallowed in strict mode
				clearcoatNormalMap: { value: null },
				^^^^^^^^^^^^^^^^^^

1 error(s), 0 warning(s)
```

Comment (donmccurdy) [reply src/renderers/shaders/ShaderLib.js]: Fixed, thanks!

**Comment by donmccurdy** [comment]
Moved the core changes to a new PR, https://github.com/mrdoob/three.js/pull/18769. Will keep the GLTFLoader portion here, to check in afterward.

**Comment by takahirox** [comment]
Do you plan to implement GLTFLoader clearcoat extension on top of the new plugin system #18484? Or do you first go without it?

**Comment by donmccurdy** [comment]
I'm willing to do either. The new plugin API is a big change, though, and I'm having trouble deciding how quickly we should move forward on that.

**Comment by takahirox** [comment]
I see, thanks. Regarding the new plugin API, let's keep discussing in #18484

**Comment by donmccurdy** [comment]
@takahirox if I polish up this PR this week, would you be OK with merging it before the plugin API?

**Comment by takahirox** [comment]
That's fine because I'm thinking to create a new PR so I don't care if this change can make the conflicts with the existing plugin PR. Then I'll make a branch after this PR gets in.

**Comment by Mugen87** [comment]
https://raw.githack.com/donmccurdy/three.js/feat-KHR_materials_clearcoat/examples/webgl_loader_gltf_extensions.html

Um, it seems the URL of the new asset is not correct since I get a 404.

**Comment by emackey** [comment]
@donmccurdy Looks like a case issue.  In your branch, rename the model and its folder from `ClearCoatTest` to `ClearcoatTest`.

**Comment by donmccurdy** [comment]
Fixed ✅

**Comment by mrdoob** [comment]
@donmccurdy feel free to merge this when you think it's ready 👌

**Pull request merged**